modal/_asgi.py

```diff
@@ -28,26 +33,32 @@ async def fetch_data_in():
             message_gen = function_io_manager.get_data_in.aio(function_call_id)
             try:
                 first_message = await asyncio.wait_for(message_gen.__anext__(), 5.0)
-            except asyncio.TimeoutError:
+            except (asyncio.TimeoutError, StopAsyncIteration):
                 if scope["type"] == "http":
-                    await messages_from_app.put({"type": "http.response.start", "status": 502})
                     await messages_from_app.put(
-                        {
-                            "type": "http.response.body",
-                            "body": b"Missing request, possibly due to cancellation or crash",
-                        }
+                        [
+                            {"type": "http.response.start", "status": 502},
+                            {
+                                "type": "http.response.body",
+                                "body": b"Missing request, possibly due to cancellation or crash",
+                            },
+                            {"type": "http.disconnect"},
+                        ]
                     )
-                    await messages_to_app.put({"type": "http.disconnect"})
                 elif scope["type"] == "websocket":
                     await messages_from_app.put(
-                        {
-                            "type": "websocket.close",
-                            "code": 1011,
-                            "reason": "Missing request, possibly due to cancellation or crash",
-                        }
+                        [
+                            {
+                                "type": "websocket.close",
+                                "code": 1011,
+                                "reason": "Missing request, possibly due to cancellation or crash",
+                            },
+                            {"type": "websocket.disconnect"},
+                        ]
                     )
-                    await messages_to_app.put({"type": "websocket.disconnect"})
-                return
+                else:
+                    logger.error(f"scope is neither http nor websocket {scope['type']})")
+                raise NoRequestReceived()
```

Comment:
This would be caught and reraised by the task.result() call in the main loop below:
```
logger.error(f"Data fetching task stopped unexpectedly: {fetch_data_in_task.exception()}")
fetch_data_in_task.result()
```
And that reraised exception would bubble all the way up to the modal function running, so it would basically error spit out a traceback and fail the function call for the user, which is probably (?) good since it's unexpected and the response will be a 502 to the client anyways

Reply:
The exception isn't actually handled in our code until after the function finishes, but this is mostly by design, since the asgi handler itself *should* raise a ClientDisconnect error before that (since we send an http.disconnect message to the asgi app).

I think this failure state isn't greatly handled by our asgi wrapper today either, since it typically leads to both us and whatever asgi framework the user uses to sending error asgi messages back to modal-http...